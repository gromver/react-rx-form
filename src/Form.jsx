/**
 *
 * Form
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Model, Form } from 'rx-form';

class FormComponent extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      form: new Form({
        model: props.model,
      }),
    };
  }

  componentWillReceiveProps(next) {
    if (next.model !== this.props.model) {
      this.cleanup();

      this.setState({ form: new Form({ model: next.model }) });
    }
  }

  componentWillUnmount() {
    this.cleanup();
  }

  // eslint-disable-next-line class-methods-use-this
  cleanup() {
    // todo destroy form
  }

  render() {
    const { render } = this.props;

    return render(this.state.form);
  }
}

FormComponent.propTypes = {
  render: PropTypes.func.isRequired,
  model: PropTypes.instanceOf(Model).isRequired,
};

export default FormComponent;
