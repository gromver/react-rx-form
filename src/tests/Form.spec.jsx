import React from 'react';
import { shallow } from 'enzyme';
import { Model, Form as RxForm } from 'rx-form';
import Form from '../Form';

describe('Test Form', () => {
  test('Form should render', async () => {
    const model = Model.object({});
    const mockFn = jest.fn((form) => <span>{JSON.stringify(form)}</span>);
    const component = shallow(<Form
      render={mockFn}
      model={model}
    />);

    expect(component).toMatchSnapshot();
    expect(mockFn).toHaveBeenCalledWith(expect.any(RxForm));
  });

  test('Form should re-render when the model prop has been changed', async () => {
    const model = Model.object({});
    const mockFn = jest.fn(() => null);
    const component = shallow(<Form
      render={mockFn}
      model={model}
    />);

    component.setProps({
      model: Model.object({}),
    });
    expect(mockFn.mock.calls[0][0]).not.toBe(mockFn.mock.calls[1][0]);
  });

  test('Form should cleanup', async () => {
    const model = Model.object({});
    const component = shallow(<Form
      render={() => null}
      model={model}
    />);

    const mockFn = jest.spyOn(component.instance(), 'cleanup');
    component.unmount();

    expect(mockFn).toHaveBeenCalled();
  });
});
