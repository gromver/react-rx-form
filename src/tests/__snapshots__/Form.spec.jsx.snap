// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form Form should render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormComponent
    model={
      Model {
        "context": Object {},
        "initialValue": Object {},
        "observable": Subject {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "observers": Array [],
          "thrownError": null,
        },
        "scenarios": Array [
          "default",
        ],
        "states": Object {},
        "type": ObjectType {
          "filter": undefined,
          "normalizeType": [Function],
          "permission": undefined,
          "properties": Object {},
          "validator": undefined,
          "validatorConfig": Object {},
        },
        "value": Object {},
      }
    }
    render={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": "{\\"model\\":{\\"initialValue\\":{},\\"value\\":{},\\"type\\":{\\"properties\\":{},\\"validatorConfig\\":{}},\\"states\\":{},\\"observable\\":{\\"_isScalar\\":false,\\"observers\\":[],\\"closed\\":false,\\"isStopped\\":false,\\"hasError\\":false,\\"thrownError\\":null},\\"scenarios\\":[\\"default\\"],\\"context\\":{}},\\"scope\\":[]}",
    },
    "ref": null,
    "rendered": "{\\"model\\":{\\"initialValue\\":{},\\"value\\":{},\\"type\\":{\\"properties\\":{},\\"validatorConfig\\":{}},\\"states\\":{},\\"observable\\":{\\"_isScalar\\":false,\\"observers\\":[],\\"closed\\":false,\\"isStopped\\":false,\\"hasError\\":false,\\"thrownError\\":null},\\"scenarios\\":[\\"default\\"],\\"context\\":{}},\\"scope\\":[]}",
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": "{\\"model\\":{\\"initialValue\\":{},\\"value\\":{},\\"type\\":{\\"properties\\":{},\\"validatorConfig\\":{}},\\"states\\":{},\\"observable\\":{\\"_isScalar\\":false,\\"observers\\":[],\\"closed\\":false,\\"isStopped\\":false,\\"hasError\\":false,\\"thrownError\\":null},\\"scenarios\\":[\\"default\\"],\\"context\\":{}},\\"scope\\":[]}",
      },
      "ref": null,
      "rendered": "{\\"model\\":{\\"initialValue\\":{},\\"value\\":{},\\"type\\":{\\"properties\\":{},\\"validatorConfig\\":{}},\\"states\\":{},\\"observable\\":{\\"_isScalar\\":false,\\"observers\\":[],\\"closed\\":false,\\"isStopped\\":false,\\"hasError\\":false,\\"thrownError\\":null},\\"scenarios\\":[\\"default\\"],\\"context\\":{}},\\"scope\\":[]}",
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
